### ISSUE TITLE
Select or write a FROST implementation

### ISSUE THREAD

**Issue by hdevalence** [penumbra-zone/penumbra#2220]
We need two relatively custom behaviors: decaf377 and randomization. Randomization is also a requirement for using FROST in Zcash, so we should check whether the ZF's implementation would be suitable.

**Comment by hdevalence** [comment]
xref https://github.com/penumbra-zone/penumbra/issues/2219#issuecomment-1480658219

**Comment by kayabaNerve** [comment]
[modular-frost](https://docs.rs/modular-frost), Serai's IETF compliant FROST implementation, was recently audited, supporting:

- Custom curves (https://github.com/serai-dex/serai/blob/develop/crypto/ciphersuite/src/ed448.rs#L51-L72 and https://github.com/serai-dex/serai/blob/develop/crypto/frost/src/curve/ed448.rs#L8-L12 to show how trivial adding a curve is)
- Offset signing (account schemes/'randomization)
- Custom challenge functions for the included Schnorr algorithm
- Custom transcript specification (including the option to use [merlin](https://docs.rs/flexible-transcript/latest/flexible_transcript/struct.MerlinTranscript.html))
- Custom signing machines entirely. This is useful as we define a `TransactionMachine` *for an entire Monero transaction* in just two rounds, with the same API as the underlying FROST machines.
- Preprocess caching
- Determination of signing set after preprocessing

It also builds on our, also audited, [dkg](https://docs.rs/dkg) crate which implemented the DKG described in the FROST paper, with an encryption scheme which is effectively ICE FROST (being designed for adversarial environments requiring full identification). 

TL;DR I believe it meets all your requirements, and then some, and is audited. I'd personally recommend it :)

---

Copied over by request. If you'd like me to remove my original post, as it's quite long and I don't want to leave it in the wrong place as a distraction, please let me know :)

**Comment by hdevalence** [comment]
@kayabaNerve thanks! I cleaned up the tracking issue already, we'll definitely check out the crate.

**Comment by hdevalence** [comment]
After chatting with @avahowell, we decided this seems like a good choice to start with!

**Comment by kayabaNerve** [comment]
:D Happy to hear it could be useful. Let me know if I can help in any way.
